### PR TITLE
udev-test: check if permitted to create block device nodes

### PR DESCRIPTION
**Pull request by abogdanenko**
Build script started by mkosi with config "mkosi.debian" fails on test stage.

It turns out that systemd-nspawn has enough privileges to create "/dev/null" character device, pass preliminary check and start running test cases, but does not have permissions to create a block device, so all test cases fail.

To help diagnose the issue, this PR adds message to show why test-udev fails.

Before:

    Assertion 'mknod(devname, mode, devnum) == 0' failed at ../src/test/test-udev.c:116, function run(). Aborting.
    Assertion 'unlink(devname) == 0' failed at ../src/test/test-udev.c:118, function run(). Aborting.

After:

    mknod() failed for '/dev/sda': Operation not permitted
    unlink('/dev/sda') failed: No such file or directory

### CODE REVIEW

**Comment by abogdanenko** [comment]
I looked into failed autopkgtests.


bionic-amd64:

```
upstream             FAIL non-zero exit status 1
```

bionic-i386:

```
upstream             FAIL timed out
```

The failed tests are not related to this PR. udev-test passes:

bionic-amd64:

```
0 errors occurred

autopkgtest [19:21:53]: test udev: -----------------------]
autopkgtest [19:21:54]: test udev:  - - - - - - - - - - results - - - - - - - - - -
udev                 PASS
```

bionic-i386:

```
0 errors occurred

autopkgtest [19:30:43]: test udev: -----------------------]
udev                 PASS
autopkgtest [19:30:43]: test udev:  - - - - - - - - - - results - - - - - - - - - -
```

File: test/udev-test.pl [hunk 1]
```diff
@@ -1554,6 +1554,17 @@ sub udev_setup {
...
         }
 
+        # check if we are permitted to create block device nodes
+        my $block_device_filename = $udev_dev . "/sda";
+        if (system("mknod", $block_device_filename, "b", "8", "0")) {
+                warn "unable to create $block_device_filename";
+                return 0;
+        }
+        else
+        {
```
Comment (poettering) [inline test/udev-test.pl]: please follow coding style, we usually place `} else {` on the same line.

File: test/udev-test.pl [hunk 2]
```diff
@@ -1554,6 +1554,17 @@ sub udev_setup {
...
+                unlink $block_device_filename;
+        }
```
Comment (poettering) [inline test/udev-test.pl]: no need to have the else block btw. The unlink can be invoked directly here, as the if block returned early. In the interest of keeping indentation levels small, let's hence get rid of the else block, and place the unlink call in place directly.

File: src/test/test-udev.c [hunk 3]
```diff
@@ -113,9 +113,15 @@ static int run(int argc, char *argv[]) {
 
                         (void) mkdir_parents_label(devname, 0755);
                         (void) sd_device_get_devnum(dev, &devnum);
-                        assert_se(mknod(devname, mode, devnum) == 0);
+                        if (mknod(devname, mode, devnum))
+                        {
+                                return log_error_errno(errno, "mknod: '%s': %m", devname);
+                        }
```
Comment (poettering) [inline src/test/test-udev.c]: please follow coding style: no {} around single line if blocks, please. (And when: place { on same line as 'if' please).

For details see https://systemd.io/CODING_STYLE

Comment (poettering) [reply src/test/test-udev.c]: Oh, and we usually check return values of syscalls with `< 0`. i.e.

```c
if (mknod(…) < 0)
        return log_error_errno(errno, …);
```

File: src/test/test-udev.c [hunk 4]
```diff
@@ -113,9 +113,15 @@ static int run(int argc, char *argv[]) {
...
                 } else {
-                        assert_se(unlink(devname) == 0);
+                        if (unlink(devname))
+                        {
+                                return log_error_errno(errno, "unlink: '%s': %m", devname);
+                        }
```
Comment (poettering) [inline src/test/test-udev.c]: as above

File: test/udev-test.pl (hunk 2 above)
Comment (abogdanenko) [reply test/udev-test.pl]: done

File: test/udev-test.pl (hunk 1 above)
Comment (abogdanenko) [reply test/udev-test.pl]: OK. I've removed the else block as advised below.

File: src/test/test-udev.c (hunk 3 above)
Comment (abogdanenko) [reply src/test/test-udev.c]: I've removed braces and updated return value checks.

File: src/test/test-udev.c (hunk 4 above)
Comment (abogdanenko) [reply src/test/test-udev.c]: done

**Comment by abogdanenko** [comment]
The reason why we see "bionic-arm64 — autopkgtest finished (failure)" is not related, see #11104 .

**Comment by abogdanenko** [comment]
Function `udev_setup`, which this PR changes, has been updated in master (see #11099), so I did a rebase.

**Comment by poettering** [comment]
thanks! lgtm!

**Pull request merged**